gateway/src/metrics.rs

```diff
@@ -0,0 +1,1177 @@
+                        ),
+                        (Ok(datum), MeasurementKind::InputCurrent) => {
+                            Sample::new(
+                                target,
+                                &metric::InputCurrent { sensor, datum },
+                            )
+                        }
+                        (Ok(datum), MeasurementKind::InputVoltage) => {
+                            Sample::new(
+                                target,
+                                &metric::InputVoltage { sensor, datum },
+                            )
+                        }
+                        (Ok(datum), MeasurementKind::Speed) => Sample::new(
+                            target,
+                            &metric::FanSpeed { sensor, datum },
+                        ),
+                        (Err(e), kind) => {
+                            let kind = match kind {
+                                MeasurementKind::Temperature => "temperature",
+                                MeasurementKind::Current => "current",
+                                MeasurementKind::Voltage => "voltage",
+                                MeasurementKind::Power => "power",
+                                MeasurementKind::InputCurrent => {
+                                    "input_current"
+                                }
+                                MeasurementKind::InputVoltage => {
+                                    "input_voltage"
+                                }
+                                MeasurementKind::Speed => "fan_speed",
+                            };
+                            let error = match e {
+                                MeasurementError::InvalidSensor => {
+                                    "invalid_sensor"
+                                }
+                                MeasurementError::NoReading => "no_reading",
+                                MeasurementError::NotPresent => "not_present",
+                                MeasurementError::DeviceError => "device_error",
+                                MeasurementError::DeviceUnavailable => {
+                                    "device_unavailable"
+                                }
+                                MeasurementError::DeviceTimeout => {
+                                    "device_timeout"
+                                }
+                                MeasurementError::DeviceOff => "device_off",
+                            };
+                            let datum = sensor_errors
+                                .entry(SensorErrorKey {
+                                    name: sensor.clone(),
+                                    kind,
+                                    error,
+                                })
+                                .or_insert(Cumulative::new(0));
+                            // TODO(eliza): perhaps we should treat this as
```

Comment:
IMO, the approach you have here is right. (It _is_ currently level-triggered, I think?)

It also occurs to me that we have the ability to emit _missing_ samples to oximeter. It's not critical, but you could consider generating those, either in addition to or instead of the error counters you already have. These are nice in that we get a reason string, which missing samples don't have, but the missing samples are inline with the non-missing samples, so you get information about them from fetching just one timeseries.

Reply:
Oh, interesting --- I wasn't aware of the concept of missing samples, that's actually much nicer. Perhaps we should use that instead!

Reply:
> IMO, the approach you have here is right. (It _is_ currently level-triggered, I think?)

Whoops, I meant to say "edge-triggered" in the comment. Presently, we will increment the error count metric on _every_ poll that the sensor is still reporting that error. I was thinking it might be better for the poll error count to only be incremented when we observe a *different* error, or observe an error after having previously seen a successful measurement.